services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/HankeServiceImpl.kt

```diff
@@ -271,39 +245,160 @@ class HankeServiceImpl(private val hankeRepository: HankeRepository,
-        addHankeYhteystietoEntitysToList(h.toteuttajat, hankeEntity, ContactType.TOTEUTTAJA)
+    private fun processCreateYT(hankeYht: HankeYhteystieto, validYT: Boolean, contactType: ContactType, userid: Int,
+            hankeEntity: HankeEntity ) {
+        if (validYT) {
+            // ... it is valid, so create a new YT
+            val hankeYhtEntity = HankeYhteystietoEntity(
+                    contactType,
+                    hankeYht.sukunimi,
+                    hankeYht.etunimi,
+                    hankeYht.email,
+                    hankeYht.puhelinnumero,
+                    hankeYht.organisaatioId,
+                    hankeYht.organisaatioNimi,
+                    hankeYht.osasto,
+
+                    userid, // createdByUserId
+                    getCurrentTimeUTCAsLocalTime(), // createdAt
+                    null,
+                    null,
+                    null, // will be set by the database
+                    hankeEntity) // reference back to parent hanke
+            hankeEntity.listOfHankeYhteystieto.add(hankeYhtEntity)
+        } else {
+            // ... missing some mandatory fields, should not have gotten here. Log it and skip it.
+            logger.error {
+                "Got a new Yhteystieto object with one or more empty mandatory fields, skipping it. HankeId ${hankeEntity.id}"
+            }
+        }
     }
 
-    internal fun addHankeYhteystietoEntitysToList(listOfHankeYhteystiedot: List<HankeYhteystieto>, hankeEntity: HankeEntity, contactType: ContactType) {
-
-        listOfHankeYhteystiedot.forEach { hankeYht ->
-            if (hankeYht.sukunimi.isNotBlank() && hankeYht.etunimi.isNotBlank() && hankeYht.email.isNotBlank()
-                    && hankeYht.puhelinnumero.isNotBlank()) {
-                val hankeYhtEntity = HankeYhteystietoEntity(
-                        contactType,
-                        hankeYht.sukunimi,
-                        hankeYht.etunimi,
-                        hankeYht.email,
-                        hankeYht.puhelinnumero,
-                        hankeYht.organisaatioId,
-                        hankeYht.organisaatioNimi,
-                        hankeYht.osasto,
-                        1, //TODO: real user , make sure you don't update by another user but only once
-                        hankeYht.createdAt?.toLocalDateTime(),
-                        null, //TODO: real user , make sure updated only if real changes? or what is the design decision?
-                        getCurrentTimeUTCAsLocalTime(),  //TODO: only if changed?  do we always want to change the date? how do we know has subobject really been updated?
-                        hankeYht.id,
-                        hankeEntity)
-
-                hankeEntity.listOfHankeYhteystieto!!.add(hankeYhtEntity)
+    private fun processUpdateYT(hankeYht: HankeYhteystieto, existingYTs: MutableMap<Int, HankeYhteystietoEntity>,
```

Comment:
We might loose some readability when we shorten Yhteystieto to YT in method name.. :)

Reply:
Will change. The "YT" in the method is a left-over; the method names were much longer earlier, and shortening the YT improved readability at that time.